packages/wrangler/src/__tests__/patch-config.test.ts

```diff
@@ -0,0 +1,880 @@
+					binding: "DB2",
+				},
+			],
+		},
+		expectedToml: dedent`
+				compatibility_date = "2022-01-12"
+				name = "test-name"
+
+				[[kv_namespaces]]
+				binding = "KV"
+
+				[[kv_namespaces]]
+				binding = "KV2"
+
+				[[kv_namespaces]]
+				binding = "KV3"
+
+				[[d1_databases]]
+				binding = "DB"
+
+				[[d1_databases]]
+				binding = "DB2"
+
+			`,
+		expectedJson: dedent`
+				{
+					"compatibility_date": "2022-01-12",
+					"name": "test-name",
+					"kv_namespaces": [
+						{
+							"binding": "KV"
+						},
+						{
+							"binding": "KV2"
+						},
+						{
+							"binding": "KV3"
+						}
+					],
+					"d1_databases": [
+						{
+							"binding": "DB"
+						},
+						{
+							"binding": "DB2"
+						}
+					]
+				}
+			`,
+	},
+];
+
+const replacingOnlyTestCases: Omit<TestCase, "additivePatch">[] = [
+	{
+		name: "edit a binding",
+		original: {},
+		replacingPatch: {
+			kv_namespaces: [
+				{
+					binding: "KV2",
+				},
+			],
+		},
+		expectedToml: dedent`
+				compatibility_date = "2022-01-12"
+				name = "test-name"
+
+				[[kv_namespaces]]
+				binding = "KV2"
+
+				`,
+		expectedJson: dedent`
+				{
+					"compatibility_date": "2022-01-12",
+					"name": "test-name",
+					"kv_namespaces": [
+						{
+							"binding": "KV2"
+						}
+					]
+				}
+			`,
+	},
+	{
+		name: "add a field to an existing binding",
+		original: {
+			kv_namespaces: [
+				{
+					binding: "KV",
+				},
+			],
+		},
+		replacingPatch: {
+			kv_namespaces: [
+				{
+					binding: "KV",
+					id: "1234",
+				},
+			],
+		},
+		expectedToml: dedent`
+				compatibility_date = "2022-01-12"
+				name = "test-name"
+
+				[[kv_namespaces]]
+				binding = "KV"
+				id = "1234"
+
+				`,
+		expectedJson: dedent`
+				{
+					"compatibility_date": "2022-01-12",
+					"name": "test-name",
+					"kv_namespaces": [
+						{
+							"binding": "KV",
+							"id": "1234"
+						}
+					]
+				}
+			`,
+	},
+	{
+		name: "delete an existing binding so that none are left",
+		original: {
+			kv_namespaces: [
+				{
+					binding: "KV",
+				},
+			],
+		},
+		replacingPatch: {
+			kv_namespaces: undefined,
+		},
+		expectedToml: dedent`
+				compatibility_date = "2022-01-12"
+				name = "test-name"
+
+				`,
+		expectedJson: dedent`
+				{
+					"compatibility_date": "2022-01-12",
+					"name": "test-name"
+				}
+			`,
+	},
+	// This one doesn't work because the jsonc-parser leaves behind a stray bracket when deleting
+	// there are possibly solutions that I am not inclined to solve right now
+	// e.g. passing in {binding: undefined} in the patch instead and cleaning up empty objects
+	// {
+	// 	name: "delete an existing binding (but some bindings of that type are still left)",
+	// 	original: {
+	// 		kv_namespaces: [
+	// 			{
+	// 				binding: "KV",
+	// 			},
+	// 			{
+	// 				binding: "KV2",
+	// 			},
+	// 		],
+	// 	},
+	// 	replacingPatch: {
+	// 		kv_namespaces: [
+	// 			{
+	// 				binding: "KV",
+	// 			},
+	// 			undefined,
```

Comment:
Does this work if you take out the `undefined`?

Reply:
no :/ I probably need to rethink the 'delete' api, but since no one is asking for it now and all of this is quite separate from the patchConfig flow that is actually used, I was just going to leave it alone for now. 

Happy to take the edit with replacement stuff out of this PR as well, and just have the additive patch if you'd rather.

Reply:
for future reference: 

the edit-with-replacement is going through each terminal? leaf? node and updating that, in order to preserve comments. that's why it needs 'undefined' in order to actually delete something. I suppose the better way to do it would be work out the diff between the patch and original, and only edit those. Will iterate on this later 👍